src/javascript.ts

```diff
@@ -102,3 +103,108 @@ export class JSHandle<T = any> {
     return this._context._delegate.handleToString(this, true /* includeType */);
   }
 }
+
+export function prepareFunctionCall<T>(
+  pageFunction: Function,
+  context: ExecutionContext,
+  args: any[],
+  toCallArgumentIfNeeded: (value: any) => { handle?: T, value?: any }): { functionText: string, values: any[], handles: T[] } {
+
+  let functionText = pageFunction.toString();
+  try {
+    new Function('(' + functionText + ')');
+  } catch (e1) {
+    // This means we might have a function shorthand. Try another
+    // time prefixing 'function '.
+    if (functionText.startsWith('async '))
+      functionText = 'async function ' + functionText.substring('async '.length);
+    else
+      functionText = 'function ' + functionText;
+    try {
+      new Function('(' + functionText  + ')');
+    } catch (e2) {
+      // We tried hard to serialize, but there's a weird beast here.
+      throw new Error('Passed function is not well-serializable!');
+    }
+  }
+
+  const guids: string[] = [];
+  const handles: T[] = [];
+  const pushHandle = (handle: T): string => {
+    const guid = platform.guid();
+    guids.push(guid);
+    handles.push(handle);
+    return guid;
+  };
+
+  const visited = new Set<any>();
+  let error: string | undefined;
+  const visit = (arg: any, depth: number): any => {
+    if (!depth) {
+      error = 'Argument nesting is too deep';
+      return;
+    }
+    if (visited.has(arg)) {
+      error = 'Argument is a circular structure';
+      return;
+    }
+    if (Array.isArray(arg)) {
+      visited.add(arg);
+      const result = [];
+      for (let i = 0; i < arg.length; ++i)
+        result.push(visit(arg[i], depth - 1));
+      visited.delete(arg);
+      return result;
+    }
+    if (arg && (typeof arg === 'object') && !(arg instanceof JSHandle)) {
+      visited.add(arg);
+      const result: any = {};
+      for (const name of Object.keys(arg))
+        result[name] = visit(arg[name], depth - 1);
+      visited.delete(arg);
+      return result;
+    }
+    if (arg && (arg instanceof JSHandle)) {
+      if (arg._context !== context)
+        throw new Error('JSHandles can be evaluated only in the context they were created!');
+      if (arg._disposed)
+        throw new Error('JSHandle is disposed!');
+    }
+    const { handle, value } = toCallArgumentIfNeeded(arg);
+    if (handle)
+      return pushHandle(handle);
+    return value;
+  };
+
+  args = args.map(arg => visit(arg, 100));
+  if (error)
+    throw new Error(error);
+
+  if (!guids.length)
+    return { functionText, values: args, handles: [] };
+
+  functionText = `(...args) => {
```

Comment:
nit: argCount, handleCount, ...argsAndHandles

Reply:
also, these variables can be accessed by the user functions. Lets call it something like `__playwright_args`.
Then later
```js
const [argCount, handleCount, ...argsAndHandles] = __playwright_args;
```